src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
  * @typedef {import('@web3-storage/gateway-lib').UnixfsContext} UnixfsContext
  */
 
+/**
+ *
+ * @param {string} s
+ * @returns {import('./bindings.js').TokenMetadata}
+ */
+function deserializeTokenMetadata (s) {
+  // TODO should this be dag-json?
+  return JSON.parse(s)
+}
+
+/**
+ *
+ * @param {import('./bindings.js').TokenMetadata} m
+ * @returns string
+ */
+function serializeTokenMetadata (m) {
+  // TODO should this be dag-json?
+  return JSON.stringify(m)
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {import('@web3-storage/gateway-lib/handlers').CID} cid
+ * @returns {Promise<import('./constants.js').RATE_LIMIT_EXCEEDED>}
+ */
+async function checkRateLimitForCID (env, cid) {
```

Comment:
This only ever uses one value from env - consider simply passing the API instance not the entire environment.